File: bun/acquire.py
#!/usr/bin/env python3 -B
"""
acquire.py: active learning. Warm start, then iterated best/rest scoring.

Local options:
    --learn.start=4    initial labels
    --learn.budget=50  rows to evaluate
    --learn.check=5    guesses to check
"""
from core import *
loadDoc(__doc__)

def acquireWithBayes(data, best, rest, row):
  n = len(best.rows) + len(rest.rows)
  return likes(rest, row, n, 2) - likes(best, row, n, 2)

def acquireWithCentroid(data, best, rest, row):
  return distx(data, row, mids(best)) - distx(data, row, mids(rest))

def warm_start(data, rows, label):
  lab = clone(data, rows[:the.learn.start])
  lab.rows.sort(key=lambda r: disty(lab, label(data, r)))
  n = int(sqrt(len(lab.rows)))
  return (lab,
          clone(data, lab.rows[:n]),
          clone(data, lab.rows[n:]),
          rows[the.learn.start:])

def dont_let_Best_grow_too_big(best, rest, lab):
  if len(best.rows) > sqrt(len(lab.rows)):
    best.rows.sort(key=lambda r: disty(lab, r))
    rest.rows.append(
      add(rest.cols, sub(best.cols, best.rows.pop())))

def acquire(data, score=acquireWithCentroid,
            label=lambda _, row: row):
  rows = data.rows[:]
  shuffle(rows)
  lab, best, rest, unlab = warm_start(data, rows[:the.few], label)
  fn = lambda row: score(lab, best, rest, row)
  for _ in range(the.learn.budget):
    if not unlab: break
    pk, *unlab = sorted(unlab, key=fn)
    add(lab, add(best, label(data, pk)))
    dont_let_Best_grow_too_big(best, rest, lab)
  lab.rows.sort(key=lambda r: disty(lab, r))
  return lab

__all__ = ["acquireWithBayes", "acquireWithCentroid", "warm_start",
           "dont_let_Best_grow_too_big", "acquire"]

File: bun/acquire.js
// acquire.js: active learning (warm start, then iterated best/rest scoring)
// (c) 2026 Tim Menzies, timm@ieee.org, MIT license
//
// likes lives in core.js (also used by bayes.js).

import {
  the, loadDoc, sqrt, flr, shuffle, clone, sub, add,
  distx, disty, mids, likes
} from "./core.js";

loadDoc(`
  --learn.start=4    initial labels
  --learn.budget=50  rows to evaluate
  --learn.check=5    guesses to check
`);

function acquireWithBayes(data, best, rest, row) {
  let n = best.rows.length + rest.rows.length;
  return likes(rest, row, n, 2) - likes(best, row, n, 2); }

let acquireWithCentroid = (data, best, rest, row) =>
  distx(data, row, mids(best)) - distx(data, row, mids(rest));

function warm_start(data, rows, label) {
  let lab = clone(data, rows.slice(0, the.learn.start));
  lab.rows.sort((a,b) => disty(lab, label(data,a)) - disty(lab, label(data,b)));
  let n = flr(sqrt(lab.rows.length));
  return [lab,
          clone(data, lab.rows.slice(0, n)),
          clone(data, lab.rows.slice(n)),
          rows.slice(the.learn.start)]; }

function dont_let_Best_grow_too_big(best, rest, lab) {
  if (best.rows.length > sqrt(lab.rows.length)) {
    best.rows.sort((a,b) => disty(lab, a) - disty(lab, b));
    let popped = best.rows.pop();
    rest.rows.push(add(rest.cols, sub(best.cols, popped))); } }

function acquire(data, score, label) {
  score = score || acquireWithCentroid;
  label = label || ((_, row) => row);
  let rows = shuffle(data.rows.slice());
  let [lab, best, rest, unlab] = warm_start(data, rows.slice(0, the.few), label);
  for (let i = 0; i < the.learn.budget; i++) {
    if (!unlab.length) break;
    unlab.sort((a,b) => score(lab,best,rest,a) - score(lab,best,rest,b));
    let pk = unlab.shift();
    add(lab, add(best, label(data, pk)));
    dont_let_Best_grow_too_big(best, rest, lab); }
  lab.rows.sort((a,b) => disty(lab, a) - disty(lab, b));
  return lab; }

export { acquireWithBayes, acquireWithCentroid, warm_start,
         dont_let_Best_grow_too_big, acquire };
